pkg/driver/aws-ebs/aws_ebs_tags_controller.go

```diff
@@ -0,0 +1,441 @@
+				continue
+			}
+			klog.Infof("Successfully updated PV annotations and tags for volume %s", volume.Name)
+		}
+	}
+	return nil
+}
+
+// updateEBSTags updates the tags of an AWS EBS volume
+func (c *EBSVolumeTagsController) updateEBSTags(pvBatch []*v1.PersistentVolume, ec2Client *ec2.EC2, resourceTags []configv1.AWSResourceTag) error {
+
+	// Merge the existing tags with new resource tags
+	mergedTags := newAndUpdatedTags(resourceTags)
+
+	// Create or update the tags
+	_, err := ec2Client.CreateTags(&ec2.CreateTagsInput{
+		Resources: pvsToResourceIDs(pvBatch),
+		Tags:      mergedTags,
+	})
+
+	if err != nil {
+		return fmt.Errorf("error creating tags for volume %v: %v", pvBatch, err)
+	}
+	return nil
+}
+
+func (c *EBSVolumeTagsController) listPersistentVolumesWithRetry(ctx context.Context) ([]*v1.PersistentVolume, error) {
+	backoff := wait.Backoff{
+		Duration: operationDelay,
+		Factor:   operationBackoffFactor,
+		Steps:    operationRetryCount,
+	}
+	var pvList []*v1.PersistentVolume
+	err := wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+		var apiError error
+		pvList, apiError = c.commonClient.KubeInformers.InformersFor("").Core().V1().PersistentVolumes().Lister().List(labels.Everything())
+		if apiError != nil {
+			klog.Errorf("error listing volumes objects: %v", apiError)
+			return false, nil
+		}
+		return true, nil
+	})
+	return pvList, err
+}
+
+func (c *EBSVolumeTagsController) updateVolumeWithRetry(ctx context.Context, pv *v1.PersistentVolume) error {
+	backoff := wait.Backoff{
+		Duration: operationDelay,
+		Factor:   operationBackoffFactor,
+		Steps:    operationRetryCount,
+	}
+	err := wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+		var apiError error
+		_, apiError = c.commonClient.KubeClient.CoreV1().PersistentVolumes().Update(ctx, pv, metav1.UpdateOptions{})
+		if apiError != nil {
+			klog.Errorf("error updating volume object %s: %v", pv.Name, apiError)
+			return false, nil
+		}
+		return true, nil
+	})
+	return err
+}
+
+func (c *EBSVolumeTagsController) handleTagUpdateFailure(batch []*v1.PersistentVolume, error error) {
+	errorMessage := fmt.Sprintf("Error updating tags for volume %v: %v", batch, error)
+	for _, pv := range batch {
+		klog.Errorf("Error updating volume %v tags: %v", pv.Name, errorMessage)
+		c.failedQueue.AddRateLimited(pv.Name)
+	}
+	var pvNames []string
+	for _, pv := range batch {
+		pvNames = append(pvNames, pv.Name)
+	}
+	// Emit a warning event for the failure
+	c.eventRecorder.Warning("EBSVolumeTagsUpdateFailed", fmt.Sprintf("Failed to update tags for batch %v: %v", pvNames, errorMessage))
+}
+
+// newAndUpdatedTags adds and update existing AWS tags with new resource tags from OpenShift infrastructure
+func newAndUpdatedTags(resourceTags []configv1.AWSResourceTag) []*ec2.Tag {
+	// Convert map back to slice of ec2.Tag
+	var tags []*ec2.Tag
+	for _, tag := range resourceTags {
```

Comment:
I am confused. The function definition says - it is supposed to "merge" tags, but all I see is, this function returns `configv1.AWSResourceTag` as `ec2.Tag` objects. I do not see any "merge" happening.

Reply:
Updated the definition, Thanks for pointing out